File: problems/problems_LCR_112/solution.go
package problemLCR_112

import (
	"encoding/json"
	"log"
	"maps"
	"slices"
	"strings"
)

var dirs = [][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

func longestIncreasingPath(matrix [][]int) (ans int) {
	m, n := len(matrix), len(matrix[0])
	dp := make([][]int, m)
	for i := range dp {
		dp[i] = make([]int, n)
	}
	idxMap := make(map[int][][2]int)
	for i, row := range matrix {
		for j, val := range row {
			idxMap[val] = append(idxMap[val], [2]int{i, j})
		}
	}
	sortedVals := slices.Sorted(maps.Keys(idxMap))
	for _, val := range sortedVals {
		for _, pos := range idxMap[val] {
			i, j := pos[0], pos[1]
			for _, dir := range dirs {
				if ni, nj := i+dir[0], j+dir[1]; ni >= 0 && ni < m && nj >= 0 && nj < n && matrix[ni][nj] < val {
					dp[i][j] = max(dp[i][j], dp[ni][nj])
				}
			}
			dp[i][j]++
			ans = max(ans, dp[i][j])
		}
	}
	return
}

func Solve(inputJsonValues string) any {
	inputValues := strings.Split(inputJsonValues, "\n")
	var matrix [][]int

	if err := json.Unmarshal([]byte(inputValues[0]), &matrix); err != nil {
		log.Fatal(err)
	}

	return longestIncreasingPath(matrix)
}

File: problems/problems_LCR_112/solution.py
from collections import defaultdict

import solution
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.longestIncreasingPath(test_input)

    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        m, n = len(matrix), len(matrix[0])
        max_length = [[0] * n for _ in range(m)]
        idx_map = defaultdict(list)
        for i, row in enumerate(matrix):
            for j, val in enumerate(row):
                idx_map[val].append((i, j))
        ans = 0
        for val in sorted(idx_map.keys()):
            for i, j in idx_map[val]:
                for d in dirs:
                    if 0 <= (ni := i + d[0]) < m and 0 <= (nj := j + d[1]) < n and matrix[ni][nj] < matrix[i][j]:
                        max_length[i][j] = max(max_length[i][j], max_length[ni][nj])
                max_length[i][j] += 1
                ans = max(ans, max_length[i][j])
        return ans
